[Spring 체스 - 2단계] 디우(김동규) 미션 제출합니다.

tco0427 opened the pull request:
안녕하세요 앨런! 디우입니다😁

이전 1단계에서 피드백 주신 내용을 반영하며 함께 상태를 제거하여 여러명이 접속했을 때 상태를 공유해서 발생할 수 있는 문제를 제거해보았습니다!!
또한 이전에 DM 드렸던 것과 같은 이유로 `game_name` 대신 별도의 AutoIncrement를 사용하는 컬럼을 추가하여 해당 컬럼을 PK로 사용하도록 변경하였습니다.
마지막으로 이번 레벨2 미션을 진행하며 `@ExceptionHandler` 를 이용하여 예외처리 로직을 별도로 분리해보았습니다!
이번 2단계 미션도 잘부탁드립니다!😃

tco0427 committed:
refactor: StrandardCharsets 클래스가 제공하는 "표준 문자 집합"에 대한 상수 이용

tco0427 committed:
fix: @AfterEach는 "private" 접근제어자이면 안되므로 수정

[AfterEach (JUnit 5.0.1 API](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/AfterEach.html)

tco0427 committed:
refactor: 실생시간에 jdbcTemplate이 변경되지 않도록 fianl 선언 추가

tco0427 committed:
refactor: ChessService가 상태(ChessGame 인스턴스 변수)를 갖지 않도록 수정

tco0427 committed:
refactor: save를 사용하는 곳이 없으므로 private 접근제어자로 변경

tco0427 committed:
refactor: Service 변경에 따른 Controller 변경

tco0427 committed:
refactor: Exception 처리 분리

Exception 처리를 따로 분리하여 처리할 수 있도록 하였습니다.
또한 HttpServletRequest의 getHeader() 메소드에 "Referer"를 전달한 값으로 redirect하도록, 즉 이전 페이지로 redirect 하도록 구현하였습니다.

tco0427 committed:
feat: 체스방 목록 조회하기

localhost 요청 시 체스방 목록을 조회할 수 있도록 기능을 구현하였습니다.

tco0427 committed:
refactor: SQL문 업데이트

tco0427 committed:
feat: 스프링 적용 이전 WebController 제거

tco0427 committed:
fix: chess.hbs 제목 변경

테스트를 위해 변경해두었던 제목을 발견하지 못하였고, 수정하였습니다.

tco0427 committed:
feat: ChessGame의 AutoIncrementKey 사용

ChessGame의 자동증가 컬럼을 PK로 사용하도록 수정하였습니다.

tco0427 committed:
docs(README): 체스방에는 고유식별값이 부여된다.

tco0427 committed:
feat: 비밀번호 입력 받는 기능 추가

tco0427 committed:
feat: 체스방 목록 버튼 추가

tco0427 committed:
feat: "재시작" 버튼 제거

tco0427 committed:
feat: 저장된 게임 시작하는 기능 추가

tco0427 committed:
feat: 최소 입력 글자수 조건 추가

tco0427 committed:
feat: 게임 삭제 기능 구현

tco0427 committed:
feat: "새로운 게임"으로 버튼이름 변경

tco0427 committed:
docs(README): 구현 목록 업데이트

tco0427 committed:
feat: 진행중인 체스방은 삭제할 수 없도록 구현

tco0427 committed:
feat: 게임시작시 비밀번호를 입력 제거

게임 시작시 비밀번호를 입력받을 필요가 없으므로 제거하였습니다.

tco0427 committed:
refactor: 움직임과 관련된 IllegalMoveException 추가

tco0427 committed:
feat: 체스방 목록에서 체스방 제목을 클릭하면 체스 게임을 이어서 진행할 수 있다.

tco0427 committed:
refactor: 동일 게임 이름 여부 검사 변경

동일한 체스 게임 이름이 존재하는지 여부를 검사하는 것을 count 집계함수를 쓰도록 수정하였습니다.

tco0427 committed:
test: Dao 변경에 따른 테스트코드 수정 및 추가

tco0427 committed:
refactor: 움직임 예외 처리시 문자열 파싱 방법 변경

쿼리스트링만 제거하고 이전 페이지로 redirect하도록 변경

tco0427 committed:
fix: 게임삭제시 비밀번호 틀리면 에러 메시지 출력

IllegalPasswordException을 사용하여 게임삭제시 비밀번호가 틀리면 에러 메시지를 출력해주도록 하였습니다.

hongbin-dev commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -25,72 +23,89 @@ public SpringController(ChessService chessService) {
-    public String start(@RequestParam("game_name") String gameName) throws UnsupportedEncodingException {
-        String encodedParam = URLEncoder.encode(gameName, "UTF-8");
-        return "redirect:/game/" + encodedParam;
+    public String start(@RequestParam("game_name") String gameName, @RequestParam("password") String password) {
+        Long id = chessService.save(gameName, password);
```
체스게임을 시작하는 요청같은데요. GetMapping으로 하신 이유가 있을까요?
클라이언트입장에서는 Get method는 조회하는걸로 인지할 수 있는데, 자원이 생성되네요. PostMapping이 자연스러울 것 같은데 어떻게 생각하시나요?

tco0427 replied:
오늘 안그래도 적절한 HTTP Method사용에 대한 강의를 들으면서 해당 부분을 잘못사용하고 있다는 것을 인지하였습니다!! 새로운 것을 생성하는 경우에는 POST 메소드가 더 적절한 것 같습니다! 감사합니다😃
(POST 메소드는 주로 새로운 리소스를 생성할 때 사용한다.)

hongbin-dev commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -25,72 +23,89 @@ public SpringController(ChessService chessService) {
+        chessService.deleteByGameNameAndPassword(state, gameName, password);
+
+        return "redirect:/";
+    }
```
`@PostMapping("/delete")`로 지정하신 이유가 있을까요? @DeleteMapping을 고려해볼 수 있지않을까요?

hongbin-dev commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -25,72 +23,89 @@ public SpringController(ChessService chessService) {
         return "chess";
     }
 
+    @PostMapping("/delete")
```
추가로 여기는 game의 식별자를 받지않는이유가 있을까요?

tco0427 replied:
handlebars를 사용하는 상태에서 DeleteMaping을 적절히 사용할 수 있는 방법을 찾지못하였고, POST를 유지하기로 앨런과 결정하였으므로 "/delete"를 유지하여 "삭제"의 의미를 나타낼 수 있도록 하겠습니다....🥲

(구현한 코드와는 별개로, @DeleteMapping("/game")과 같이 DeleteMapping을 통해 '삭제'라는 의미를 나타내고, URL은 자원을 나타낼 수 있도록 하는 것이 적절하다고 생각합니다!)

hongbin-dev replied:
url을 보고 어떤 리소스를 삭제하려는지 파악은 어려운 것 같아요.
만약 식별자로 삭제한다면 `/game/:id/delete` 이런 네이밍을 고려해볼 수 있을 것 같아요.

tco0427 replied:
API를 설계할 때 "행위"는 http 메소드로 지정(행위는 URL이 아닌 Method로 표현하기)하고, 자원을 명사로써 path로 지정하라 라는 내용의 `체스 미션 피드백` 수업을 들었습니다!!
그런 의미에서 `/game/:id/delete` 보다 `/game/:id` 를 URL로써 사용하고, @DeleteMapping 혹은 메소드명을 통해서 삭제의 의미를 나타내는 건 어떤가요..??

hongbin-dev replied:
> 그런 의미에서 /game/:id/delete 보다 /game/:id 를 URL로써 사용하고, @DeleteMapping 혹은 메소드명을 통해서 삭제의 의미를 나타내는 건 어떤가요..??

네 좋습니다! 웹기반 컨트롤러에서 delete method로 요청받고, redirect 시킬 때 GET으로 요청되지않는 이슈가 있으니 참고해주세요. (delete -> redirect -> delete)

hongbin-dev commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -25,72 +23,89 @@ public SpringController(ChessService chessService) {
+    @PostMapping("/delete")
+    public String delete(@RequestParam(value = "game_name", required = false) String gameName,
+                         @RequestParam(value = "password", required = false) String password) {
+        State state = chessService.findStateByGameNameAndPassword(gameName, password);
+        chessService.deleteByGameNameAndPassword(state, gameName, password);
+
+        return "redirect:/";
```
```suggestion
        chessService.deleteBy(id, password);

        return "redirect:/";
```
또 컨트롤러 내에서 비즈니스의 플로우를 작성하지마시고, 서비스에 위임하는게 좋아보입니다.
다른 컨트롤러도 적용해주세요~

tco0427 replied:
컨트롤러는 "gameName과 password를 통해서 삭제를 한다" 라는 것에만 관심이 있지, "어떻게(how)" 삭제를 하는지는 알 필요가 없다고 생각합니다. 즉, 말씀해주신대로 비즈니스 흐름(어떤 메소드를 호출하는지 그리고 어떤 순서로 호출하는지)에 대해서는 알 필요가 없고, 이를 서비스에 위임하는게 적절해 보입니다!! 수정해보도록 하겠습니다!

(이전까지는 deleteByGameNameAndPassword()와 같이 메소드명을 작성하였는데, 이렇게 메소드명이 길어지는 경우 deleteBy(gameName, password)와 같은 네이밍으로 메소드명도 짧으면서 의미를 확실히 전달할 수 있다는 생각이 드네요!! 이 부분도 한 번 반영해보도록 하겠습니다!😃)

hongbin-dev commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -6,44 +6,70 @@
-        String sql = "insert into chessgame (game_name, turn) values (?, ?)";
-        jdbcTemplate.update(sql, chessGameDto.getGameName(), chessGameDto.getTurn());
+    public Long save(ChessGameDto chessGameDto, String password) {
+        validateDuplicateGameName(chessGameDto.getGameName());
```
DAO: 하나의 쿼리를 보내는 책임
Service: 트랜잭션단위의 비즈니스 흐름을 책임

중복이름을 제어가 필요한건 비즈니스 영역이기 때문에 Service로 이동하는게 좋아보이네요.

tco0427 replied:
동의합니다. DAO는 Data Access Object가 의미하는 것과 같이 DB에 접근하고 쿼리를 날려 결과를 받아오는데에만 집중하면 된다고 생각합니다. 따라서 말씀해주신 것과 같이 트랜잭션단위의 비즈니스 흐름을 책임지는 Service 쪽으로 옮기는 것이 적절한 것 같습니다!
"체스 게임을 생성할 때 이미 동일한 게임이름이 존재하는지 유무를 검사"하는 부분을 Service쪽으로 이동해보았습니다.

hongbin-dev commented on src/main/java/chess/dao/ChessGameDao.java:
```diff
@@ -6,44 +6,70 @@
-    public void update(ChessGameDto chessGameDto) {
-        String sql = "update chessgame set turn = ? where game_name = ?";
-        jdbcTemplate.update(sql, chessGameDto.getTurn(), chessGameDto.getGameName());
+    private void validateDuplicateGameName(String gameName) {
+        String sql = "select count(*) from CHESSGAME where game_name = ?";
```
```suggestion
    private void validateDuplicateGameName(String gameName) {
        String sql = "select * from CHESSGAME where game_name = ? LIMIT 1";
```
limit 1을 사용하면 모든 테이블을 탐색하지 않을 수도 있겠네요

tco0427 replied:
LIMIT을 사용하여 하나의 row가 검색되면 더 이상 탐색하지 않음으로써 모든 테이블의 row를 탐색하지 않아도 되어 성능개선이 있을 수 있을 것 같습니다! (특히 row가 10,000개 이상 100,000개 이상과 같이 정말 많이지는 경우 그 이점을 더욱 명확히 느낄 수 있을 것 같습니다!) 개선해보도록 하겠습니다😃

tco0427 replied:
jdbcTemplate.queryForObject()는 조회 결과가 없는 경우 `EmptyResultDataAccessException` 을 발생시킵니다.
따라서 위의 코멘트 변경과 함께 해당 쿼리문을 수정하여 다음과 같이 코드를 작성할 수 있었습니다.

```java
private void validateDuplicateGameName(String gameName) {
    try {
        chessGameDao.findOneByGameName(gameName);
        throw new IllegalGameProgressException("동일한 이름의 게임이 이미 존재합니다.");
    } catch (EmptyResultDataAccessException ignored) {
    }
}
```

즉, 우리가 원하는 실행흐름(정상흐름)이 `EmptyResulstDataAccessException`이기 때문에 위와 같이 코드를 작성하게 되었습니다. 그런데 try-catch구문에서 오히려 예외가 발생할 때 정상흐름이 된다는게 어색하게 느껴졌습니다.
1. 실무에서도 위와같은 코드가 생길 때가 종종 있는지 궁금합니다!
2. 위와 같은 코드에서 발생할 수 있는 문제점이 있을까요? 만약 있다면 어떤 것이 있는지 궁금합니다!
저는 우선 가독성 측면에서 단점이 있다고 생각했습니다. (이 코드를 처음보는 사람이 이해하기 위해서는 많이 시간이 들어가지 않을까?)

hongbin-dev replied:
디우 이 부분은 DM으로 답변드렸습니다! exists 쿼리를 이용해볼 수 있을 것 같아요

tco0427 replied:
감사합니다!!!😃

hongbin-dev commented on src/main/java/chess/dao/PieceDao.java:
```diff
@@ -12,22 +12,21 @@
 @Repository
 public class PieceDao {
-    private JdbcTemplate jdbcTemplate;
+    private final JdbcTemplate jdbcTemplate;
```
👍

hongbin-dev commented on src/main/java/chess/exception/ExistGameException.java:
```diff
@@ -0,0 +1,7 @@
+package chess.exception;
+
+public class ExistGameException extends IllegalArgumentException {
+    public ExistGameException() {
+        super("동일한 이름의 게임이 이미 존재합니다");
+    }
+}
```
커스텀예외를 등록하신 이유가 있을까요?

tco0427 replied:
모든 IllegalArgumentException이 논리적으로 동일한 예외라는 생각이 들지 않았습니다.
예를 들어, 크게는 "동일한 게임 이름 존재"와 "체스게임에서 움직임 예외"는 동일하지 않다는 생각을 하였습니다.
그래서 이 예외들을 구분해주기 위해서 커스텀예외를 등록하였습니다..!
(또 예외 메시지도 가능하면 해당 커스텀 예외 안으로 위임해주고 싶었습니다..😅)

hongbin-dev replied:
답변 감사합니다. 또 커스텀예외를 사용했을 때 단점도 있을까요?

tco0427 replied:
자바에서 표준예외를 잘 제공하고 있기 때문에, 커스텀 에외를 남용했을 때 문제가 발생할 수 있을 것 같습니다.
인자값이 부적절한 경우에 던지는 예외인 `IllegalArgumentException` 이나 작업을 수행하기에 부적합한 상태인 경우 던지는 예외인 `IllegalStateException` 또 `UnsupportedOperationException` 등 자바에서 제공하는 우리에게 익숙한 표준 예외를 사용함으로써 충분히 예외를 표현할 수 있습니다.
따라서 이런 표준예외를 적절하게 사용하지 않고, 커스텀 예외를 계속해서 만들게 되면 우리가 "직접" 관리 해줘야 할 커스텀 예외 클래스가 지나치게 많아질 수 있다고 생각합니다.

다음의 글을 참고해보았습니다!! [custom exception을 언제 써야 할까?](https://tecoble.techcourse.co.kr/post/2020-08-17-custom-exception/)

hongbin-dev commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -5,95 +5,102 @@
 
     private final ChessGameDao chessGameDao;
     private final PieceDao pieceDao;
-    private ChessGame chessGame;
```
👍

hongbin-dev commented on src/main/java/chess/controller/ExceptionController.java:
```diff
@@ -0,0 +1,44 @@
+    public String handleMove(RedirectAttributes redirectAttributes, IllegalArgumentException e, HttpServletRequest request) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        String referer = request.getHeader("Referer");
+        String[] split = referer.split("\\?");
+
+        return "redirect:" + split[0];
```
이쪽 코드는 이전 페이지로 이동한다고 볼 수 있을까요?

tco0427 replied:
네 맞습니다! 이전 페이지로 이동하지만 뒤에 붙는 쿼리 스트링을 제거해주지 않으면 계속해서 중첩되는 문제가 있어 쿼리스트링을 제거해주기 위해 referer.split("\\?"); 가 필요해졌습니다.

hongbin-dev commented on src/main/java/chess/controller/ExceptionController.java:
```diff
@@ -0,0 +1,44 @@
+        return "redirect:" + split[0];
+    }
+
+    @ExceptionHandler(IllegalPasswordException.class)
+    public String handleEmptyResult(RedirectAttributes redirectAttributes, IllegalPasswordException e) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        return "redirect:/";
+    }
+
+    @ExceptionHandler(IllegalDeleteException.class)
+    public String handleDelete(RedirectAttributes redirectAttributes, IllegalDeleteException e) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        return "redirect:/";
+    }
+
+    @ExceptionHandler(ExistGameException.class)
+    public String handleExistGame(RedirectAttributes redirectAttributes, ExistGameException e) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        return "redirect:/";
+    }
```
크게 다른 행위를 하지않는다면 하나로 묶을 수 있지않을까요?

tco0427 replied:
아하! 이런 관점에서 앨런이 커스텀 예외를 등록한 이유를 여쭤본 것 같습니다!
이렇게 보면 "게임 움직임에 관련된 예외"와 그렇지 않은 예외, 즉 예외에 따른 처리로 분리를 할 수 있을 것 같습니다.
그리고 이렇게 되면 굳이 IllegalPasswordException, IllegalDeleteException, ExistGameException 이라는 커스텀 예외를 만들 필요가 없게 되네요😅
개선해보겠습니다!

tco0427 replied:
`IllegalPasswordException`, `IllegalDeleteException`, `ExistGameException` 이 세가지 예외를 묶은 하나의 또 다른 예외(IllegalGameProgressException)를 만들었습니다.
단순히 `IllegalArgumentException`을 사용하게 되면 `IllegalMoveExcpetion` 이 `IllegalArgumentException`을 상속받으므로 컴파일 에러가 발생하기 때문에..

hongbin-dev commented on src/main/java/chess/controller/ExceptionController.java:
```diff
@@ -0,0 +1,44 @@
+        return "redirect:" + split[0];
+    }
+
+    @ExceptionHandler(IllegalPasswordException.class)
+    public String handleEmptyResult(RedirectAttributes redirectAttributes, IllegalPasswordException e) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        return "redirect:/";
+    }
```
서버에서 발생한 예외메시지를 클라이언트에 그대로 전달하고 있는데요. 이 부분은 좀 더 고민해보면 좋을 것 같습니다.

tco0427 replied:
지금의 경우 클라이언트에 보여줄 예외 메시지가 어떠한 이유로 변경된다면 서버의 코드(Exception의 예외메시지 부분)를 변경해야하는 문제가 있을 수 있을 것 같습니다. 따라서 서버에서 ResponseEntity를 사용하여 예외에 따른 적절한 상태코드를 front에 넘겨주고 front에서 적절하게 요청에 따른 예외 메시지를 보여줄 수 있을 것 같습니다..!!
그런데..현재와 같이 Handlerbars 사용시에는 어떻게 예외와 그에 대한 메시지를 분리해줄 수 있을지 모르겠습니다...ㅠㅠ

hongbin-dev replied:
Handlerbars를 사용하는경우는 상태코드를 활용하기 어렵지만, 
도메인에서 발생한 예외메시지와 사용자에게 전달할 예외메시지는 달라야할 것 같습니다! 중요한 정보가 노출될 가능성이 있다고 생각해요

hongbin-dev commented on src/main/java/chess/controller/SpringController.java:
```diff
@@ -25,72 +23,89 @@ public SpringController(ChessService chessService) {
 
     @GetMapping("/")
-    public String home() {
+    public String home(Model model, @RequestParam(value = "error", required = false) String error) {
```
현재 `SpringController`로 되어있는데요. 이쪽에 책임이 너무 많은 것 같아서요.
크게 HomeController와 GameController로 나눠볼 수 있을 것 같아요

tco0427 replied:
게임진행(ex. 기물 움직임)과 관련된 사용자 요청을 받아서 처리하는 쪽(GameController)과 전체적인 게임 관리(ex. 게임 삭제, 게임 시작 등)(HomeController)로 분리하여볼 수 있을 것 같습니다! 개선해보도록 하겠습니다~_~

hongbin-dev reviewed (changes requested):
안녕하세요 디우~
피드백 적용과 2단계미션 잘 해주셨군요 👍
질문들과 코멘트 남겼으니 확인해주시고요. 궁금하신점은 편하게 DM 해주세요!

tco0427 committed:
refactor: GetMapping -> PostMapping

"/start" 요청시 새로운 리소스(새로운 체스게임)를 생성하므로 Post 메소드가 더 적절하다고 판단하여 수정해주었습니다.

tco0427 committed:
refactor: 비즈니스 플로우 제거

비즈니스 플로우를 제거하고, 서비스에 위임하는 방향으로 수정하였습니다.

tco0427 committed:
feat: 누락된 init.sql 업데이트

tco0427 committed:
refactor: 예외 공통처리

동일한 예외처리를 하는 예외들을 하나로 묶었습니다.

tco0427 committed:
refactor: 컨트롤러 분리

게임 진행과 관련된 요청 처리를 GameController로 그 이외의 요청에 대한 처리는 HomeController로 책임을 분리하였습니다.

tco0427 committed:
refactor: 쿼리문 개선

- 게임이 이미 존재하는지를 확인하는 것은 "비즈니스 흐름을 책임"지는 Service에서 하는것이 적절하므로 Service단으로 옮겼습니다.
- LIMIT을 이용하여 하나의 row가 발견되면 더이상 탐색하지 않음으로써 모든 테이블의 row를 검사하지 않도록 개선해보았습니다.

tco0427 committed:
refactor: EXISTS와 서브쿼리 이용

EXISTS와 서브쿼리를 이용하여 `EmptyResultDataAccessException`을 피하고 Boolean값을 받아오도록 수정하였습니다.

hongbin-dev commented on src/main/java/chess/controller/ExceptionController.java:
```diff
@@ -21,22 +19,8 @@ public String handleMove(RedirectAttributes redirectAttributes, IllegalArgumentE
-
-    @ExceptionHandler(ExistGameException.class)
-    public String handleExistGame(RedirectAttributes redirectAttributes, ExistGameException e) {
+    @ExceptionHandler(IllegalGameProgressException.class)
+    public String handle(RedirectAttributes redirectAttributes, IllegalGameProgressException e) {
```
👍

hongbin-dev commented on src/main/java/chess/controller/ExceptionController.java:
```diff
@@ -0,0 +1,28 @@
+    public String handleMove(RedirectAttributes redirectAttributes, IllegalArgumentException e, HttpServletRequest request) {
+        redirectAttributes.addAttribute("error", e.getMessage());
+
+        String referer = request.getHeader("Referer");
+        String[] split = referer.split("\\?");
+
+        return "redirect:" + split[0];
```
```suggestion
        String referer = request.getHeader("Referer");
        String previousPath = URI.create(referer)
            .getPath();

        return "redirect:" + previousPath;
```
기존 자바패키지에 제공하는 기능을 활용해보는건 어떨까요?

tco0427 replied:
URI.create() 을 통해서 인자로 전달한 문자열로 URI 객체를 생성하고, getPath() 를 이용해서 해당 URI의 디코딩된 경로만을 반환받음으로써 쿼리스트링의 중첩을 막을 수 있겠군요! 감사합니다!!

hongbin-dev commented on src/main/java/chess/service/ChessService.java:
```diff
@@ -5,95 +5,114 @@
 
-    public List<String> move(String moveCommand) {
-        Command command = Command.from(moveCommand);
+    public Long save(String gameName, String password) {
```
디우가 생각하는 서비스레이어는 어떤 역할이라고 생각하시나요~

tco0427 replied:
비즈니스 플로우에 대한 책임을 가진다고 생각합니다. 예를 들어, ChessGame을 생성하고, progress() 메시지를 보내는 행위 혹은 이전에 코멘트 주셨던 것과 같이 DAO를 사용해서 State(상태)를 조회해오고 이를 사용해 삭제를 하는 등의 비즈니스 흐름에 대한 책임을 가지고 있다고 생각합니다. 즉, DAO 조회 결과등을 가공 그리고 도메인을 사용하여 비즈니스 로직의 '흐름'을 제어하는 책임을 가지고 있다고 이야기할 수 있을 것 같습니다.
또한 저의 코드에서 `validateDuplicateGameName()` 과 같은 비즈니스 로직도 Service의 역할이라고 생각합니다!

정리하면 다음과 같을 것 같이 정리할 수 있을 것 같습니다!
Controller: 사용자 요청을 받고, 그에 따른 적절한 Service 호출 및 응답
Service: 비즈니스 로직 처리(도메인의 비즈니스 로직 처리와는 차이가 있다고 생각합니다. 예를 들어, "게임진행(progress)"이라는 것은 도메인의 책임이지만, "중복된 이름이 있는지 검증한 후 게임을 진행시키고 이를 DB에 저장한다."(도메인을 활용한 비즈니스 흐름 제어)라는 비즈니스 책임은 Service라고 이야기 할 수 있을 것 같습니다.
DAO: DB 접근에 대한 책임 (단순히 쿼리를 날리고, 그에 대한 결과를 반환)

현재 제가 생각하는  서비스 계층은 위와 같은데 지적해주시거나 보충해주시면 너무 감사할 것 같습니다!!🙏
(저번에 포비께서 다른 크루에게 같은 질문을 할 때, 생각해보았던 것인데, 아직 제 생각에 확신은 없는 것 같아 미션을 진행하며 사용한 Controller, Service, DAO를 기반으로 정리해보았습니다.😃)

hongbin-dev replied:
정리 감사합니다!

> Service: 비즈니스 로직 처리(도메인의 비즈니스 로직 처리와는 차이가 있다고 생각합니다. 예를 들어, "게임진행(progress)"이라는 것은 도메인의 책임이지만, "중복된 이름이 있는지 검증한 후 게임을 진행시키고 이를 DB에 저장한다."(도메인을 활용한 비즈니스 흐름 제어)라는 비즈니스 책임은 Service라고 이야기 할 수 있을 것 같습니다.

정리하면 서비스는 도메인로직의 묶음이라고 생각할 수 있겠군요. 

> "중복된 이름이 있는지 검증한 후 게임을 진행시키고 이를 DB에 저장한다."

이런내용이면 서비스레이어에서 영속성관련된 비즈니스 룰을 책임질 수 있겠고, 서비스에서 트랜잭션을 책임지기도 하겠네요.
적다보니 서비스레이어에 트랜잭션을 추가하면 좋겠군요!

tco0427 replied:
넵!! 감사합니다!! 보충해서 학습하도록 하겠습니다 😃

hongbin-dev reviewed (changes requested):
안녕하세요 디우!
피드백 잘 반영해주셨군요 💯
몇 가지 질문과 코멘트 남겼으니 확인해주세요~

tco0427 committed:
refactor: URI 객체 사용

URI 객체를 사용하여 쿼리스트링이 중첩되는 문제를 해결하도록 수정하였습니다.
(자바패키지에서 제공하는 기능 활용)

tco0427 committed:
refactor: IllegalMoveException 사용

tco0427 committed:
refactor: 도메인 예외 메시지와 뷰 예외 메시지 분리

도메인에서 발생한 예외메시지와 뷰에서 발생하는 예외 메시지를 분리하였습니다.

hongbin-dev reviewed (approved):
안녕하세요 디우!
피드백 반영 잘 해주셨군요 💯💯
스프링 체스 미션 고생많으셨습니다. 이만 머지할게요
